### PR TITLE
doc: document mds journaling

### PR DESCRIPTION
**Pull request by vshankar**

### CODE REVIEW

**Comment by vshankar** [comment]
jenkins render docs

**Comment by batrick** [comment]
````
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/jenkins-build/build/workspace/ceph-pr-docs/build-doc/virtualenv/lib/python3.5/site-packages/sphinx/util/docutils.py:311: RemovedInSphinx30Warning: function based directive support is now deprecated. Use class based directive instead.
  RemovedInSphinx30Warning)
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/cephfs/mds-journaling.rst: WARNING: document isn't included in any toctree
done
````

File: doc/cephfs/mds-journaling.rst [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+CephFS uses journaling for couple of reasons:
+
+#. Consistency: On an MDS failover, the journal events can be replayed to reach a
+                consistent filesystem state. Also, metadata operations that require
+                multiple updates to the backing store need to be journaled for crash
+                consistency (along with other consistency mechanisms such as locking,
+                etc..).
+
+#. Performance: Journal updates are (mostly) sequential, hence updates to journals
+                are fast.
```
Comment (batrick) [inline doc/cephfs/mds-journaling.rst]: Updates can also be batched into a single write.

File: doc/cephfs/mds-journaling.rst [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
```
Comment (batrick) [inline doc/cephfs/mds-journaling.rst]: A large journal also indirectly helps with getting the cache warm during failover.

**Comment by vshankar** [comment]
jenkins render docs

**Comment by vshankar** [comment]
jenkins render docs

File: doc/cephfs/mds-journaling.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+CephFS Metadata Pool
+====================
+
+CephFS uses a separate (metadata) pool for managing file metadata (inodes and
+dentries) in a Ceph File System. The metadata pool has all the information about
+files in a Ceph File System including the File System heirarchy. Additionally,
+CephFS maintains meta information related to other entities in a filesystem
+such as file system journals, open file table, session map, etc.
+
+This document describes how Ceph Metadata Server use and rely on journaling.
```
Comment (fullerdj) [inline doc/cephfs/mds-journaling.rst]: nit: s/Server/Servers

File: doc/cephfs/mds-journaling.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+Each active metadata server maintains its own journal in the metadata pool. Journals
+are striped over multiple objects. Journal entries which are not required (deemed as
+old) are trimmed by the metadata server.
+
+==============
+Journal Events
+==============
+
+Apart from journaling file system metadata updates, CephFS journals various other events
+such as client session info, directory import/export state to name a few. These events
```
Comment (fullerdj) [inline doc/cephfs/mds-journaling.rst]: nit: s/,/and/

File: doc/cephfs/mds-journaling.rst [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+are used by the metadata sever to reestablish correct state as required, e.g., Ceph MDS
+tries to reconnects clients on restart when journal events get replayed and a specific
```
Comment (fullerdj) [inline doc/cephfs/mds-journaling.rst]: nit: s/reconnects/reconnect

File: doc/cephfs/mds-journaling.rst [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+event type in the journal specifies that a client entity type has a session with the MDS
+before it was restarted.
```
Comment (fullerdj) [inline doc/cephfs/mds-journaling.rst]: This sentence is a little hard to follow. Maybe separating the events into a numbered list would make it easier.

Comment (vshankar) [reply doc/cephfs/mds-journaling.rst]: yeh -- that's what is planned for current (doc) sprint.

**Comment by vshankar** [comment]
jenkins render docs

File: doc/cephfs/mds-journaling.rst [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+=====================
+
+CephFS metadata servers stream a journal of metadata events into RADOS in the metadata
+pool prior to executing a file system operation. Active MDS daemon(s) manage metadata
+for files and directories in CephFS.
+
+CephFS uses journaling for couple of reasons:
+
+#. Consistency: On an MDS failover, the journal events can be replayed to reach a
+   consistent filesystem state. Also, metadata operations that require multiple
```
Comment (rishabh-d-dave) [inline doc/cephfs/mds-journaling.rst]: nit: s/filesystem/file system

File: doc/cephfs/mds-journaling.rst [hunk 8]
```diff
@@ -0,0 +1,59 @@
+.. _mds-journaling:
+
+====================
+CephFS Metadata Pool
+====================
+
+CephFS uses a separate (metadata) pool for managing file metadata (inodes and
+dentries) in a Ceph File System. The metadata pool has all the information about
+files in a Ceph File System including the File System heirarchy. Additionally,
+CephFS maintains meta information related to other entities in a filesystem
```
Comment (rishabh-d-dave) [inline doc/cephfs/mds-journaling.rst]: nit: s/filesystem/file system

**Comment by vshankar** [comment]
fixed and updated

File: doc/cephfs/mds-journaling.rst [hunk 9]
```diff
@@ -0,0 +1,59 @@
+.. _mds-journaling:
+
+====================
+CephFS Metadata Pool
+====================
+
+CephFS uses a separate (metadata) pool for managing file metadata (inodes and
+dentries) in a Ceph File System. The metadata pool has all the information about
+files in a Ceph File System including the File System heirarchy. Additionally,
```
Comment (rishabh-d-dave) [inline doc/cephfs/mds-journaling.rst]: s/heirarchy/hierarchy

Comment (vshankar) [reply doc/cephfs/mds-journaling.rst]: done

**Comment by vshankar** [comment]
jenkins render docs

File: doc/cephfs/mds-journaling.rst [hunk 10]
```diff
@@ -0,0 +1,59 @@
+.. _mds-journaling:
+
+====================
```
Comment (batrick) [inline doc/cephfs/mds-journaling.rst]: The header isn't quite right, it becomes the title for the page: https://docs.ceph.com/ceph-prs/30396/cephfs/mds-journaling/#cephfs-mds-journaling

Title should be "MDS Journaling"?

Comment (vshankar) [reply doc/cephfs/mds-journaling.rst]: ah! yes -- fixed

**Comment by vshankar** [comment]
jenkins render docs

**Comment by vshankar** [comment]
> Doc render available at http://docs.ceph.com/ceph-prs/30396/

hmmm... that didn't completely fix the header. I'll recheck.

**Comment by vshankar** [comment]
> > Doc render available at http://docs.ceph.com/ceph-prs/30396/
> 
> hmmm... that didn't completely fix the header. I'll recheck.

sorry -- the doc looks fine (something was cached). Please take a look.
